ledger/acctupdates.go

```diff
@@ -1511,37 +1538,54 @@ func (au *accountUpdates) generateCatchpoint(committedRound basics.Round, label
 	default:
 		chunkExecutionDuration = shortChunkExecutionDuration
 	}
-	for more {
-		stepCtx, stepCancelFunction := context.WithTimeout(au.ctx, chunkExecutionDuration)
-		writeStepStartTime := time.Now()
-		more, err = catchpointWriter.WriteStep(stepCtx)
-		// accumulate the actual time we've spent writing in this step.
-		catchpointGenerationStats.CPUTime += uint64(time.Now().Sub(writeStepStartTime).Nanoseconds())
-		stepCancelFunction()
-		if more && err == nil {
-			// we just wrote some data, but there is more to be written.
-			// go to sleep for while.
-			select {
-			case <-time.After(100 * time.Millisecond):
-			case <-au.ctx.Done():
-				retryCatchpointCreation = true
+
+	var catchpointWriter *catchpointWriter
+	err = au.dbs.rdb.Atomic(func(ctx context.Context, tx *sql.Tx) (err error) {
+		catchpointWriter = makeCatchpointWriter(au.ctx, absCatchpointFileName, tx, committedRound, committedRoundDigest, label)
+		for more {
+			stepCtx, stepCancelFunction := context.WithTimeout(au.ctx, chunkExecutionDuration)
+			writeStepStartTime := time.Now()
+			more, err = catchpointWriter.WriteStep(stepCtx)
+			// accumulate the actual time we've spent writing in this step.
+			catchpointGenerationStats.CPUTime += uint64(time.Now().Sub(writeStepStartTime).Nanoseconds())
+			stepCancelFunction()
+			if more && err == nil {
+				// we just wrote some data, but there is more to be written.
+				// go to sleep for while.
+				// before going to sleep, extend the transaction timeout so that we won't get warnings:
+				db.ResetTransactionWarnDeadline(ctx, tx, time.Now().Add(1*time.Second))
+				select {
+				case <-time.After(100 * time.Millisecond):
+				case <-au.ctx.Done():
+					retryCatchpointCreation = true
+					err2 := catchpointWriter.Abort()
+					if err2 != nil {
+						return fmt.Errorf("error removing catchpoint file : %v", err2)
+					}
+					return nil
+				case <-au.catchpointSlowWriting:
+					chunkExecutionDuration = longChunkExecutionDuration
+				}
+			}
+			if err != nil {
+				err = fmt.Errorf("unable to create catchpoint : %v", err)
 				err2 := catchpointWriter.Abort()
 				if err2 != nil {
 					au.log.Warnf("accountUpdates: generateCatchpoint: error removing catchpoint file : %v", err2)
 				}
 				return
-			case <-au.catchpointSlowWriting:
-				chunkExecutionDuration = longChunkExecutionDuration
 			}
 		}
-		if err != nil {
-			au.log.Warnf("accountUpdates: generateCatchpoint: unable to create catchpoint : %v", err)
-			err2 := catchpointWriter.Abort()
-			if err2 != nil {
-				au.log.Warnf("accountUpdates: generateCatchpoint: error removing catchpoint file : %v", err2)
-			}
-			return
-		}
+		return
+	})
+
+	if err != nil {
+		au.log.Warnf("accountUpdates: generateCatchpoint: %v", err)
+		return
+	}
+	if catchpointWriter == nil {
```

Comment:
It shouldn't. The `db.Atomic` is not supposed to return with nil unless the callback function was called.
However, since there could "thoererically" be a bug in the `Atomic` function, I figured it would be safer to add this no-op check.